README.md
# Energy Wiring Diagram Generator for UK TIMES

This code generates wiring diagrams for the energy flows in UK TIMES. 

i.e., it shows natural gas being connected to imports and to electricity generation and to heating

See http://decc.github.io/times-wiring-diagram/ for example output


Relies on: 

1. the excellent [Graphviz](http://graphviz.org) visualisation library
2. the excellent [svg-pan-zoom](https://github.com/ariutta/svg-pan-zoom) javascript library.
3. the excellent [d3](http://d3js.org) visualisation library

&copy; 2015 Tom Counsell http://tom.counsell.org

## Usage

You need to have installed:

1. Ruby
2. Graphviz

You need to have run the TIMES scenario, and found the resulting gdx file. This is usually in VEDA_FE/GAMS_WrkTIMES/GamsSave folder, with the same name as the scenario that has been run.

Then you run:

    ruby update-wiring-diagram.rb <GAMS-GDX-FILE.gdx> <description>

Where description is anything you like. The version of the TIMES database being tested is a good thing to say.

## Notes

This diagram only includes commodities whose unit is PJ.

This dagram combines processes and commodities whose name only differs by a number at the end.

## Hacking

Authoritative copy at http://github.com/decc/times-wiring-diagram

Please report issues there.

## TODO

1. A way of searching for a particular node then highlighting it and zooming to it
2. A way of following links to or from a node

## License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.



CHANGES.md
# Changes

## 2015-03-23

* Now draws charts left to right
* Now fills an area with roughly A4 landscape ratio
* Now aligns initial sources of energy on the left
* Now aligns final uses of energy on the right

## 2015-03-22

* Can now click on a node to highlight it and its flows.

## 2015-03-20 First public release

* First public release.


gdx.rb
require 'csv'

class Gdx
  attr_accessor :gdx_filename
  

  def initialize(gdx_filename)
    @gdx_filename = gdx_filename
    @fraction_of_year_for_timeslice = {}
  end

  def symbol(symbol)
    CSV.new(`gdxdump #{gdx_filename} Symb=#{symbol} Format=csv`, headers: :true, converters: :all, header_converters: :symbol).to_a.map(&:to_hash)
  end

  def load_time_slices
    fraction_of_year_for_timeslice = {}
    symbol(:G_YRFR).each do |row|
      fraction_of_year_for_timeslice[row[:ts]] = row[:val]
    end
    return fraction_of_year_for_timeslice
  end

  def com_units
    return @com_units if @com_units
    @com_units = {}
    symbol(:COM_UNIT).each do |commodity|
      @com_units[commodity[:com]] = commodity[:units_com].to_sym
    end
    @com_units
  end

  def commodity_is_energy?(commodity)
    com_units[commodity] == :PJ
  end

  def simplify_flow_name(flow_name)
    flow_name[/^(.*?)\d*$/,1]
  end

  def load_flows(year)
    flows = Hash.new { |hash, key| hash[key] = 0 } # Hash with default of zero for missing elements
    symbol(:F_IN).each do |flow|
      next unless flow[:allyear] = year
      from = simplify_flow_name(flow[:c])
      next unless commodity_is_energy?(from)
      to = simplify_flow_name(flow[:p])
      amount = flow[:val]
      next if amount == 'Eps'
      flows[[from, to]] += amount
    end
    symbol(:F_OUT).each do |flow|
      next unless flow[:allyear] = year
      from = simplify_flow_name(flow[:p])
      to = simplify_flow_name(flow[:c])
      next unless commodity_is_energy?(to)
      amount = flow[:val]
      next if amount == 'Eps'
      flows[[from, to]] += amount
    end
    flows
  end

  def map_flows_to_d3_format(flows)
    nodes = Hash.new { |hash, key| hash[key] = hash.size }
    links = flows.map do |flow, value|
      {
        source: nodes[flow[0]],
        target: nodes[flow[1]],
        value: value
      }
    end
    nodes = nodes.to_a.sort_by { |a| a[1] }.map { |name, index| { name: name } }
    p nodes.length
    p links.length
    { nodes: nodes, links: links }
  end

end

#puts gdx.map_flows_to_d3_format(gdx.load_flows(2010)).to_json


update-wiring-diagram.rb
require_relative 'gdx'

unless ARGV[0] && File.exist?(ARGV[0])
  puts "Usage:"
  puts "ruby update-wiring-diagram.rb <GAMS-GDX-FILE.gdx> <description>"
  puts "where the gdx file is usual found in the VEDA_FE/GAMS_WrkTIMES/GamsSave folder, with the same name as the scenario that has been run"
  puts "n.b.: this means you have to run a scenario with the TIMES model before attempting this"
  puts "<description> is added to the output, good things to put here are the version number of the TIMES model"
  exit
end

title = "Energy wiring diagram for TIMES #{ARGV[1]}"

# First we load the gdx file
puts "Opening #{ARGV[0]}"
gdx = Gdx.new(ARGV[0])

# Then we extract the flows
# The choice of 2010 is, I think arbitrary
puts "Extracting flows"
flows = gdx.load_flows(2010)

# Create the Node objects
puts "Analysing nodes"
class Node
  attr_accessor :in, :out, :name
  def initialize(name)
    @name = name
    @in, @out = [], []
  end
  def primary_source
    @in.empty? && !@out.empty?
  end
  def final_demand
    @out.empty? && !@in.empty?
  end
end

nodes = Hash.new { |hash, key| hash[key] = Node.new(key) }

# We do this so we can put all the sources of energy and final demands
# on the same horizontal row in the final diagram
puts "Finding primary sources of energy and final demands for energy"
flows.each do |flow, value|
  h = nodes[flow[0].gsub(/[^a-zA-Z0-9]+/,'')]
  t = nodes[flow[1].gsub(/[^a-zA-Z0-9]+/,'')]
  next if t == h
  h.out.push(t)
  t.in.push(h) 
end

# Then we open up a Graphviz input file:
puts "Writing Graphviz input file"
File.open('wiring.gv', 'w') do |f|
  # The Graphviz format is described here:
  # http://www.graphviz.org/
  f.puts "digraph wiring {" # We have a directed graph (or at least, in a sensible energy system we should"
  f.puts 'graph [rankdir="LR", ratio=0.7]' # We want the result to look square-ish
  f.puts 'node [id="\N"]' # We want the result to look square-ish
  f.puts 'edge [id="\T_\H"]' # We want the result to look square-ish
  f.puts "{rank=source; #{nodes.values.select(&:primary_source).map(&:name).join(' ')} }"
  f.puts "{rank=sink; #{nodes.values.select(&:final_demand).map(&:name).join(' ')} }"

  flows.each do |flow, value| # We write out each flow in the Graphviz format
    # We have to make sure that the node names don't include anything but letters and digits
    f.puts "#{flow[0].gsub(/[^a-zA-Z0-9]+/,'')} -> #{flow[1].gsub(/[^a-zA-Z0-9]+/,'')};"
  end
  
  f.puts "}"
end

# Now we process the Graphviz input file into svg
# -T svg means write the result as svg
# -o flows.svg gives the output filename
puts "Processing graphviz file"
puts `dot -o wiring.svg -T svg wiring.gv`

# Now we open the svg file
puts "Loading resulting svg"
svg = IO.readlines('wiring.svg')

# Now we drop all the gumph at the head of the file until
# we hit the first bit of drawing
svg.shift until svg.first =~ /^<g/


# Now we embed the svg in a template
html = <<END
<html>
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
<meta charset='utf-8'>
<title>#{title}</title>
<style>
  h1 {
    width: 100%;
    text-align: center;
    size: 10px;
    margin: 0;
    padding: 0;
  }
  g.node {
    pointer-events: all;
  }
  .highlight ellipse {
    fill: #ff0;
    stroke: #f00;
    stroke-width: 4px;
  }
  .highlight path {
    stroke: #f00;
    stroke-width: 4px;
  }
</style>
<script src='d3.min.js'></script>
<script src='svg-pan-zoom.min.js'></script>
<body>
<h1>#{title}</h1>
<svg id='nodeplot'>
#{svg.join}
<script>
var margin = {top: 20, right: 20, bottom: 30, left: 50},
    width = window.innerWidth - margin.left - margin.right,
    height = window.innerHeight - margin.top - margin.bottom;

d3.select('#nodeplot')
  .attr('width', width)
  .attr('height', height);

d3.selectAll(".edge")
  .attr("class", function(d) { return 'edge '+(this.id.replace('_', ' ')); });

d3.selectAll(".node")
  .attr("class", function(d) { return 'node '+(this.id); });

d3.selectAll("g.node").on('click',function(d) {
    if( d3.select(this).classed('highlight')) {
      d3.selectAll("."+this.id).classed('highlight', false);
    } else {
      d3.selectAll("."+this.id).classed("highlight", true);
    }
    d3.event.preventDefault();
    });

var panZoomTiger = svgPanZoom('#nodeplot', {
  controlIconsEnabled: true,
  zoomEnabled: true,
  maxZoom: 50
});
</script>
</html>
END

# Now we write that out to a file
puts "Writing webpage"
File.open('index.html','w') { |f| f.puts html }

# And we are done
puts "Done"



update-flow-diagram.rb
require_relative 'gdx'

unless ARGV[0] && File.exist?(ARGV[0])
  puts "Usage:"
  puts "ruby update-flow-diagram.rb <GAMS-GDX-FILE.gdx> <year> <description>"
  puts "where the gdx file is usual found in the VEDA_FE/GAMS_WrkTIMES/GamsSave folder, with the same name as the scenario that has been run"
  puts "n.b.: this means you have to run a scenario with the TIMES model before attempting this"
  puts "<year> is the year for which you want the flow diagram drawn"
  puts "<description> is added to the output, good things to put here are the version number of the TIMES model"
  exit
end

year = ARGV[1].to_i
title = "#{year} Energy flow diagram for TIMES #{ARGV[2]}"

threshold = 10 # Do not show flows with less energy (in PJ) than this
pen_thickness_per_pj = 0.01 # Scale the thickness of lines in proportion to PJ

# First we load the gdx file
puts "Opening #{ARGV[0]}"
gdx = Gdx.new(ARGV[0])

# Then we extract the flows
# The choice of 2010 is, I think arbitrary
puts "Extracting flows"
flows = gdx.load_flows(year)

# Create the Node objects
puts "Analysing nodes"
class Node
  attr_accessor :in, :out, :name
  attr_accessor :in_size, :out_size

  def initialize(name)
    @name = name
    @in, @out = [], []
    @in_size, @out_size = 0.0, 0.0
  end
  def primary_source
    @in.empty? && !@out.empty?
  end
  def final_demand
    @out.empty? && !@in.empty?
  end
  def size
    [@in_size, @out_size].max
  end
end

nodes = Hash.new { |hash, key| hash[key] = Node.new(key) }

# We do this so we can put all the sources of energy and final demands
# on the same horizontal row in the final diagram
puts "Finding primary sources of energy and final demands for energy"
flows.each do |flow, value|
  h = nodes[flow[0].gsub(/[^a-zA-Z0-9]+/,'')]
  t = nodes[flow[1].gsub(/[^a-zA-Z0-9]+/,'')]
  next if t == h
  h.out.push(t)
  h.out_size += value
  t.in.push(h) 
  t.in_size += value
end

# Then we open up a Graphviz input file:
puts "Writing Graphviz input file"
File.open('flow.gv', 'w') do |f|
  # The Graphviz format is described here:
  # http://www.graphviz.org/
  f.puts "digraph flows {" # We have a directed graph (or at least, in a sensible energy system we should"
  f.puts 'graph [rankdir="LR", ratio=0.7, arrowhead="none"]' # We want the result to look square-ish
  f.puts 'node [shape="box", id="\N"]' # We want the result to look square-ish
  f.puts 'edge [id="\T_\H"]' # We want the result to look square-ish
  f.puts "{rank=source; #{nodes.values.select(&:primary_source).map(&:name).join(' ')} }"
  f.puts "{rank=sink; #{nodes.values.select(&:final_demand).map(&:name).join(' ')} }"

  nodes.values.each do |n|
    f.puts "#{n.name} [height=#{(n.size*pen_thickness_per_pj/50).round}];"
  end

  flows.each do |flow, value| # We write out each flow in the Graphviz format
    next if value.abs < threshold # Skip the small flows
    # We have to make sure that the node names don't include anything but letters and digits
    f.puts "#{flow[0].gsub(/[^a-zA-Z0-9]+/,'')} -> #{flow[1].gsub(/[^a-zA-Z0-9]+/,'')} [weight=#{value.round}, penwidth=#{(value*pen_thickness_per_pj).round}];"
  end
  
  f.puts "}"
end

# Now we process the Graphviz input file into svg
# -T svg means write the result as svg
# -o flows.svg gives the output filename
puts "Processing graphviz file"
puts `dot -o flow.svg -T svg flow.gv`

# Now we open the svg file
puts "Loading resulting svg"
svg = IO.readlines('flow.svg')

# Now we drop all the gumph at the head of the file until
# we hit the first bit of drawing
svg.shift until svg.first =~ /^<g/
svg.delete_if { |line| line =~ /^<polygon fill="black/ } # some polygons get in the way

# Now we embed the svg in a template
html = <<END
<html>
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
<meta charset='utf-8'>
<title>#{title}</title>
<style>
  h1 {
    width: 100%;
    text-align: center;
    size: 10px;
    margin: 0;
    padding: 0;
  }
  g.node {
    pointer-events: all;
  }
  .highlight polygon {
    fill: #ff0;
    stroke: #f00;
    stroke-width: 4px;
  }
  .highlight path {
    stroke: #f00;
  }
</style>
<script src='d3.min.js'></script>
<script src='svg-pan-zoom.min.js'></script>
<body>
<h1>#{title}</h1>
<svg id='nodeplot'>
#{svg.join}
<script>
var margin = {top: 20, right: 20, bottom: 30, left: 50},
    width = window.innerWidth - margin.left - margin.right,
    height = window.innerHeight - margin.top - margin.bottom;

d3.select('#nodeplot')
  .attr('width', width)
  .attr('height', height);

d3.selectAll(".edge")
  .attr("class", function(d) { return 'edge '+(this.id.replace('_', ' ')); });

d3.selectAll(".node")
  .attr("class", function(d) { return 'node '+(this.id); });

d3.selectAll("g.node").on('click',function(d) {
    if( d3.select(this).classed('highlight')) {
      d3.selectAll("."+this.id).classed('highlight', false);
    } else {
      d3.selectAll("."+this.id).classed("highlight", true);
    }
    d3.event.preventDefault();
    });

var panZoomTiger = svgPanZoom('#nodeplot', {
  controlIconsEnabled: true,
  zoomEnabled: true,
  maxZoom: 50
});
</script>
</html>
END

# Now we write that out to a file
puts "Writing webpage"
File.open('flow.html','w') { |f| f.puts html }

# And we are done
puts "Done"

